File: frontend/src/app/models/broker.ts
//
// Date: 2/10/2018
// Author(s): Spicer Matthews (spicer@options.cafe)
// Copyright: 2018 Cloudmanic Labs, LLC. All rights reserved.
//

import { BrokerAccount } from './broker-account';

export class Broker {

  public Id: number;
  public Name: string;
  public Status: string;
  public DisplayName: string;
  public BrokerAccounts: BrokerAccount[];
  public SettingsActiveBrokerAccount: BrokerAccount;
  
  //
  // Construct.
  //
  constructor(
    Id: number,
    Name: string,
    DisplayName: string,    
    Status: string,   
    BrokerAccounts: BrokerAccount[],
  ){
    this.Id = Id;
    this.Name = Name;
    this.DisplayName = DisplayName;
    this.Status = Status;    
    this.BrokerAccounts = BrokerAccounts;
  }

  //
  // Json to Object.
  //
  public static fromJson(json: Object): Broker 
  {
    let o = new Broker(json["id"], json["name"], json["display_name"], json["status"], []);
    return o;
  }

  //
  // Build object for emitting to the app.
  //
  public static buildForEmit(data) : Broker[]  {

    let result = [];

    if(! data)
    {
      return result;      
    }

    for(let i = 0; i < data.length; i++)
    {
      // Add in the positions
      let accounts = [];
      
      for(let k = 0; k < data[i].broker_accounts.length; k++)
      {
        accounts.push(new BrokerAccount(
          data[i].broker_accounts[k].id,
          data[i].broker_accounts[k].name,
          data[i].broker_accounts[k].broker_id,  
          data[i].broker_accounts[k].account_number,           
          data[i].broker_accounts[k].stock_commission, 
          data[i].broker_accounts[k].stock_min,
          data[i].broker_accounts[k].option_commission,
          data[i].broker_accounts[k].option_single_min,
          data[i].broker_accounts[k].option_multi_leg_min,
          data[i].broker_accounts[k].option_base                         
        ));
      }

      result.push(new Broker(
        data[i].id,
        data[i].name,
        data[i].display_name,         
        data[i].status,       
        accounts
       ));
    }

    return result; 
  }  
}

/* End File */

File: backend/cmd/actions/broker.go
//
// Date: 2018-11-08
// Author: Spicer Matthews (spicer@cloudmanic.com)
// Last Modified by: Spicer Matthews
// Last Modified: 2018-11-08
// Copyright: 2017 Cloudmanic Labs, LLC. All rights reserved.
//

package actions

import (
	"encoding/json"
	"fmt"

	"github.com/cloudmanic/app.options.cafe/backend/brokers"
	"github.com/cloudmanic/app.options.cafe/backend/brokers/tradier"
	"github.com/cloudmanic/app.options.cafe/backend/library/helpers"
	"github.com/cloudmanic/app.options.cafe/backend/models"
)

//
// Dump the json response for a user's order history
//
// go run main.go --cmd="broker-json-history" --user_id=1 --broker_account_id=2 > orderhistory.json
//
func GetJsonBrokerHistory(db *models.DB, userId int, brokerAccountId int) {

	var brokerApi brokers.Api

	// Get user.
	user, err := db.GetUserById(uint(userId))

	if err != nil {
		panic(err)
	}

	// Get broker account
	brokerAccount, err := db.GetBrokerAccountByIdUserId(uint(brokerAccountId), user.Id)

	if err != nil {
		panic(err)
	}

	// Get broker
	broker, err := db.GetBrokerById(brokerAccount.BrokerId)

	// Decrypt the access token
	accessToken, err := helpers.Decrypt(broker.AccessToken)

	if err != nil {
		panic(err)
	}

	// Figure out which broker connection to setup.
	switch broker.Name {

	case "Tradier":
		brokerApi = &tradier.Api{ApiKey: accessToken, DB: db, Sandbox: false}

	case "Tradier Sandbox":
		brokerApi = &tradier.Api{ApiKey: accessToken, DB: db, Sandbox: true}

	default:
		panic("Unknown Broker : " + broker.Name + " (" + user.Email + ")")

	}

	// Make API call to get the JSON.
	orders, err := brokerApi.GetAllOrders()

	if err != nil {
		panic(err)
	}

	// Convert to json
	json, err := json.Marshal(orders)

	if err != nil {
		panic(err)
	}

	// Return JSON
	fmt.Println(string(json))
}

/* End File */
